Source/rotation/rotation_sources.cpp

```diff
@@ -0,0 +1,489 @@
+#include "Castro.H"
+#include "Castro_F.H"
+#include "Castro_util.H"
+#ifdef HYBRID_MOMENTUM
+#include "hybrid.H"
+#endif
+
+void
+Castro::rsrc(const Box& bx,
+             Array4<Real const> const& rot,
+             Array4<Real const> const& uold,
+             Array4<Real> const& source, 
+             const Real dt) {
+
+  GpuArray<Real, 3> center;
+  ca_get_center(center.begin());
+
+  GeometryData geomdata = geom.data();
+
+  amrex::ParallelFor(bx,
+  [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+  {
+
+    Real Sr[3] = {};
+    Real src[NSRC] = {};
+
+    // Temporary array for seeing what the new state would be if the update were applied here.
+
+    Real snew[NUM_STATE] = {};
+
+    GpuArray<Real, 3> loc;
+    position(i, j, k, geomdata, loc);
+
+    for (int dir = 0; dir < AMREX_SPACEDIM; ++dir) {
+      loc[dir] -= center[dir];
+    }
+
+    Real rho = uold(i,j,k,URHO);
+    Real rhoInv = 1.0_rt / rho;
+
+    for (int n = 0; n < NUM_STATE; n++) {
+      snew[n] = uold(i,j,k,n);
+    }
+
+    Real old_ke = 0.5_rt * (snew[UMX] * snew[UMX] + snew[UMY] * snew[UMY] + snew[UMZ] * snew[UMZ]) * rhoInv;
+
+    for (int n = 0; n < 3; n++) {
+      Sr[n] = rho * rot(i,j,k,n);
+    }
+
+    src[UMX] = Sr[0];
+    src[UMY] = Sr[1];
+    src[UMZ] = Sr[2];
+
+    snew[UMX] += dt * src[UMX];
+    snew[UMY] += dt * src[UMY];
+    snew[UMZ] += dt * src[UMZ];
+
+#ifdef HYBRID_MOMENTUM
+    if (state_in_rotating_frame == 1) {
+      GpuArray<Real, 3> linear_source;
+      linear_source[0] = src[UMR];
```

Comment:
In `set_hybrid_momentum_source`, the input `linear_source` is the source to the linear momenta UMX, UMY, and UMZ, and the output `hybrid_source` is the source to UMR, UML, and UMP. See construct_old_gravity_source. This is different from the interface in Fortran.

Reply:
I think I've fixed this now